Added differences to JavaScript regular expression

bobnil opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description
Added a section about the differences to JavaScript regular expression. 
<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->
The implied anchors was mentioned in the text, but easy to miss.
Nothing was mentioned about the need to escape hyphen in a character class.

By having all known differences mentioned under one section makes it easier to find.

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

The standard mentions the implied anchors in step 5 of 4.10.5.3.6, but I can't find a mention about the need to escape hyphen. Tested in Chrome, Edge, Brave, and Firefox.


### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

bobnil committed:
Added differences to JavaScript regular expression

bobnil committed:
Changed markdown character for list

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Josh-Cena commented on files/en-us/web/html/attributes/pattern/index.md:
```diff
@@ -148,6 +148,11 @@ When a control has a `pattern` attribute, the `title` attribute, if used, must d
+### Differences to JavaScript regular expression
+
+- The expression matches against the entire value, not just any subset, as if a `^(?:` were implied at the start of the pattern and `)$` at the end.
+- In a character class, a hyphen `-` must be escaped as `\-`. Example: to match digits 0-9 and hyphen `[0-9-]*` is valid in JavaScript but must be writen as `[0-9\-]*`.
```
This is simply because patterns use the `v` flag. You get the same error if you write `/[0-9-]*/v`.

chrisdavidmills replied:
Would be good to explain that, certainly.

chrisdavidmills committed:
Merge branch 'main' into patch-1

chrisdavidmills commented on files/en-us/web/html/attributes/pattern/index.md:
```diff
@@ -148,6 +148,11 @@ When a control has a `pattern` attribute, the `title` attribute, if used, must d
 
 While `title`s are used by some browsers to populate error messaging, because browsers sometimes also show the title as text on hover, it therefore shows in non-error situations, so be careful not to word titles as if an error has occurred.
 
+### Differences to JavaScript regular expression
```
```suggestion
### Differences to JavaScript regular expressions
```

chrisdavidmills replied:
Also, this entire section could do with better placement and thinking about repetition:

1. It really belongs further up the page, inside the "Overview" section. It is currently a subsection of "Examples", but it is not an example.
2. On the same subject, "Accessibility concerns" shouldn't be a subsection of "Examples" either. I know that's not your doing, but maybe you could fix that too while you're here, by promoting it to a level 2 heading?
3. The first bullet below is already covered in the "Constraint validation" section. Maybe remove that information from there, and then move this subsection so it is just below the "Constraint validation" subsection, as a follow-on?

bobnil replied:
I moved the information to be inside "Constraint validation" and rewrote it since the v flag changes how character classes are interpreted and more characters than hyphen must be escaped.

I also changed the heading of "Accessibility concerns" to level 2.

But I'm sorry to say that I don't know how to push the changes to this request.

chrisdavidmills reviewed (changes requested):
Thanks for the contribution @bobnil. I've left you a few comments to consider.

bobnil committed:
Updated explanation of how the expression is compiled.

Added information that the regular expression is compiled with the v flag and that it changes how character classes are interpreted.

Also promoted the heading of Accessibility concerns to level 2, at the request from chrisdavidmills.

Josh-Cena commented on files/en-us/web/html/attributes/pattern/index.md:
```diff
@@ -32,6 +32,8 @@ When including a `pattern`, provide a description of the pattern in visible text
 If the input's value is not the empty string and the value does not match the entire regular expression, there is a constraint violation reported by the {{domxref('ValidityState')}} object's {{domxref('ValidityState.patternMismatch','patternMismatch')}} property being `true`.
 The pattern's regular expression, when matched against the value, must have its start anchored to the start of the string and its end anchored to the end of the string, which is slightly different from JavaScript regular expressions: in the case of pattern attribute, we are matching against the entire value, not just any subset, as if a `^(?:` were implied at the start of the pattern and `)$` at  ...
 
+The pattern's regular expression is compiled with the `v` flag ([`unicodeSets`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicodeSets) mode), which changes how character classes are interpreted. In addition to `]` and `\`, the following characters must be escaped if they represent literal characters: `(`, `)`, `[`, `{`, `}`, `/`, `-`, `|`. Example: to match digits 0-9 and hyphen  ...
```
The text above states that "the `'u'` flag is specified". You would need to update that too. This would also partially address https://github.com/mdn/content/issues/26907 (barring browser compatibility)
